google/cloud/internal/win32/parse_service_account_p12_file.cc

```diff
@@ -0,0 +1,318 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef _WIN32
+#include "google/cloud/internal/parse_service_account_p12_file.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/internal/win32/win32_helpers.h"
+#include "absl/types/span.h"
+#include <fstream>
+#include <type_traits>
+#include <vector>
+#include <Windows.h>
+#include <wincrypt.h>
+
+namespace google {
+namespace cloud {
+namespace oauth2_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+using ::google::cloud::internal::FormatWin32Errors;
+using ::google::cloud::internal::InvalidArgumentError;
+
+namespace {
+
+struct HCertStoreDeleter {
+  void operator()(HCERTSTORE store) const { CertCloseStore(store, 0); }
+};
+
+using UniqueCertStore =
+    std::unique_ptr<std::remove_pointer_t<HCERTSTORE>, HCertStoreDeleter>;
+
+struct CertContextDeleter {
+  void operator()(PCCERT_CONTEXT cert) const {
+    CertFreeCertificateContext(cert);
+  }
+};
+
+using UniqueCertContext =
+    std::unique_ptr<std::remove_pointer_t<PCCERT_CONTEXT>, CertContextDeleter>;
+
+/// A subclass of `std::unique_ptr` that has a get() method that returns a
+/// native integer type. This is needed because some Windows API handles are
+/// declared as native integers instead of pointers.
+template <typename T, typename Deleter>
+class UniquePtrReinterpretProxy : private std::unique_ptr<void, Deleter> {
+  static_assert(sizeof(T) == sizeof(void*),
+                "T must be the same size as a pointer");
+
+ public:
+  explicit UniquePtrReinterpretProxy(T ptr)
+      : std::unique_ptr<void, Deleter>(reinterpret_cast<void*>(ptr), {}) {}
+
+  T get() const {
+    return reinterpret_cast<T>(this->std::unique_ptr<void, Deleter>::get());
+  }
+};
+
+struct HCryptProvDeleter {
+  void operator()(void* key) const {
+    CryptReleaseContext(reinterpret_cast<HCRYPTPROV>(key), 0);
+  }
+};
+
+struct HCryptKeyDeleter {
+  void operator()(void* key) const {
+    CryptDestroyKey(reinterpret_cast<HCRYPTKEY>(key));
+  }
+};
+
+using UniqueCryptProv =
+    UniquePtrReinterpretProxy<HCRYPTPROV, HCryptProvDeleter>;
+
+using UniqueCryptKey = UniquePtrReinterpretProxy<HCRYPTKEY, HCryptKeyDeleter>;
+
+StatusOr<UniqueCertStore> OpenP12File(std::string const& source) {
+  // Read the PKCS#12 file into memory.
+  std::vector<char> data;
```

Comment:
I wonder if this should be `std::vector<BYTE> data` and then we can save a `reinterpret_cast`?

Reply:
Changed.